docs/docs/developers/guides/smart_contracts/how_to_communicate_cross_chain.md

```diff
@@ -0,0 +1,196 @@
+---
+
+This guide shows you how to implement cross-chain communication between Ethereum (L1) and Aztec (L2) contracts using portal contracts.
+
+## Prerequisites
+
+- An Aztec contract project set up with `aztec-nr` dependency
+- Understanding of Aztec L1/L2 architecture
+- Access to Ethereum development environment for L1 contracts
+- Deployed portal contract on L1 (see [token bridge tutorial](../../tutorials/js_tutorials/token_bridge.md))
+
+## Send messages from L1 to L2
+
+### Send a message from your L1 portal contract
+
+Use the `Inbox` contract to send messages from L1 to L2. Call `sendL2Message` with these parameters:
+
+| Parameter   | Type     | Description |
+|-------------|----------|-------------|
+| `actor` | `L2Actor` | Your L2 contract address and rollup version |
+| `contentHash` | `bytes32` | Hash of your message content (use `Hash.sha256ToField`) |
+| `secretHash` | `bytes32` | Hash of a secret for message consumption |
+
+In your Solidity contract:
+
+```solidity
+import {IInbox} from "@aztec/core/interfaces/messagebridge/IInbox.sol";
+import {DataStructures} from "@aztec/core/libraries/DataStructures.sol";
+import {Hash} from "@aztec/core/libraries/crypto/Hash.sol";
+
+// ... initialize inbox, get rollupVersion from rollup contract ...
+
+DataStructures.L2Actor memory actor = DataStructures.L2Actor(l2ContractAddress, rollupVersion);
+
+// Hash your message content with a unique function signature
+bytes32 contentHash = Hash.sha256ToField(
+    abi.encodeWithSignature("your_action_name(uint256,address)", param1, param2)
+);
+
+// Send the message
+(bytes32 key, uint256 index) = inbox.sendL2Message(actor, contentHash, secretHash);
+```
+
+### Consume the message in your L2 contract
+
+To consume a message coming from L1, use the `consume_l1_to_l2_message` function within the context:
+
+- The `content_hash` must match the hash that was sent from L1
+- The `secret` is the pre-image of the `secretHash` sent from L1
+- The `sender` is the L1 portal contract address
+- The `message_leaf_index` helps the RPC find the correct message
+- If the content or secret doesn't match, the transaction will revert
+- "Consuming" a message pushes a nullifier to prevent double-spending
+
+```rust
+#[public]
+fn consume_message_from_l1(
+    secret: Field,
+    message_leaf_index: Field,
+    // your function parameters
+) {
+    // Recreate the same content hash as on L1
+    let content_hash = /* compute your content hash */;
+    
+    // Consume the L1 message
+    context.consume_l1_to_l2_message(
+        content_hash,
+        secret,
+        portal_address, // Your L1 portal contract address
+        message_leaf_index
+    );
+    
+    // Execute your contract logic here
+}
+```
+
+## Send messages from L2 to L1
+
+### Send a message from your L2 contract
+
+Use `message_portal` in your `context` to send messages from L2 to L1:
+
+```rust
+#[public]
```

Comment:
Might be good to mention that this can be done from a public or a private function